omp/matrix/csr_kernels.cpp

```diff
@@ -42,12 +42,133 @@ namespace omp {
 namespace csr {
 
 
+/**
+ * Computes the begin offsets into A and B for the specific diagonal
+ *
+ * @param diagonal  the diagonal to search
+ * @param end_row_offsets  the ending of row offsets of A
+ * @param a_len  the length of A (the number of rows)
+ * @param b_len  the length of B (the number of stored elements)
+ *
+ * @return a pair which contains (x, y) coordinate where diagonal intersects the
+ * merge path
+ */
+template <typename IndexType>
+inline std::pair<IndexType, IndexType> merge_path_search(
+    const IndexType diagonal, const IndexType* end_row_offsets,
+    const IndexType a_len, const IndexType b_len)
+{
+    auto x_min = std::max(diagonal - b_len, zero<IndexType>());
+    auto x_max = std::min(diagonal, a_len);
+
+    while (x_min < x_max) {
+        auto x_pivot = x_min + ((x_max - x_min) / 2);
+        if (end_row_offsets[x_pivot] <= (diagonal - x_pivot - 1)) {
+            x_min = x_pivot + 1;  // Contract range up A (down B)
+        } else {
+            x_max = x_pivot;  // Contract range down A (up B)
+        }
+    }
+
+    return std::make_pair(std::min(x_min, a_len), diagonal - x_min);
+}
+
+
 template <typename MatrixValueType, typename InputValueType,
-          typename OutputValueType, typename IndexType>
-void spmv(std::shared_ptr<const OmpExecutor> exec,
-          const matrix::Csr<MatrixValueType, IndexType>* a,
-          const matrix::Dense<InputValueType>* b,
-          matrix::Dense<OutputValueType>* c)
+          typename OutputValueType, typename IndexType, typename AlphaOp,
+          typename BetaOp>
+void merge_spmv(std::shared_ptr<const OmpExecutor> exec,
+                const matrix::Csr<MatrixValueType, IndexType>* a,
+                const matrix::Dense<InputValueType>* b,
+                matrix::Dense<OutputValueType>* c, AlphaOp alpha_op,
+                BetaOp beta_op)
+{
+    using arithmetic_type =
+        highest_precision<MatrixValueType, InputValueType, OutputValueType>;
+
+    auto row_ptrs = a->get_const_row_ptrs();
+    auto col_idxs = a->get_const_col_idxs();
+
+    const auto a_vals =
+        acc::helper::build_const_rrm_accessor<arithmetic_type>(a);
+    const auto b_vals =
+        acc::helper::build_const_rrm_accessor<arithmetic_type>(b);
+    auto c_vals = acc::helper::build_rrm_accessor<arithmetic_type>(c);
+
+    // Merge-SpMV variables
+    const auto num_rows = static_cast<IndexType>(a->get_size()[0]);
+    const auto nnz = static_cast<IndexType>(a->get_num_stored_elements());
+    const auto num_threads = static_cast<IndexType>(omp_get_max_threads());
+    // Merge list A: row end offsets
+    const IndexType* row_end_offsets = row_ptrs + 1;
```

Comment:
is the suffix `_offsets` necessary? How you define it, I would just call it `row_ends` or maybe `row_end_idxs`.

Reply:
I think offsets matches the meaning not idxs.

Reply:
I use row_end_ptrs to match the same term we use in CSR